src/lib.rs

```diff
@@ -191,10 +191,11 @@ fn do_init(
 
     let deps = to_extern(db, api, querier);
     let mut instance = cache.get_instance(&code_id, deps, gas_limit)?;
-    let res = call_init_raw(&mut instance, params, msg)?;
+    // We only check this result after reporting gas usage and returning the instance into the cache.
+    let res = call_init_raw(&mut instance, params, msg);
     *gas_used = gas_limit - instance.get_gas();
     cache.store_instance(&code_id, instance);
-    Ok(res)
+    Ok(res?)
 }
```

Comment:
This change means that even if we run out of gas, or experience any other error, we will still report how much gas was used before the error happened

Reply:
This also happens to fix an issue where wasmer instances were not recycled if a VmError happened during execution

Reply:
I explicitly did not want to reuse instances where a VmError occured, as they can leave the wasm memory in an invalid state (eg. when a panic is thrown, it doesn't unwrap and release memory).

Good to set `gas_used` in any case. How about:

```rust
    let res = call_init_raw(&mut instance, params, msg);
    *gas_used = gas_limit - instance.get_gas();
    if res.is_ok() {
        cache.store_instance(&code_id, instance);
    }
    res
```

(Do you need `Ok(res?)` seems redundant unless there is magic error mapping happening there, in which case I think `res.map_err(|e| ...)` is clearer even if a bit longer.

Reply:
Isn't the memory reset when calling into the contract again? huh... wouldn't that mean that we ALWAYS have to recreate the `Instance` because static variables will be persisted across calls to the contract? (!!!)

>unless there is magic error mapping happening there

There is. it's equivalent to `res.map_err(|err| err.into())`

Reply:
> There is. it's equivalent to `res.map_err(|err| err.into())`

I prefer to use this instead of `Ok(res?)` as it is more clear. But that is my opinion.

Reply:
> Isn't the memory reset when calling into the contract again? huh... wouldn't that mean that we ALWAYS have to recreate the Instance because static variables will be persisted across calls to the contract? (!!!)

Caching is off by default in wasmd now. The idea is a big speedup, opt-in via config.
We do not have mutable static variables in any code, just `const` globals. Memory is inside an instance, so "reusing it" is just like calling the instance twice in a row, which will persist any memory (and have slow leaks if we have some memory leak)

Reply:
caching of instances after erroers in execution is disabled in 09d755c